src/Orchard.Web/Modules/Orchard.OutputCache/Filters/CacheByRoleFilter.cs

```diff
@@ -0,0 +1,34 @@
+﻿using System;
+using System.Linq;
+using System.Text;
+using System.Web.Mvc;
+using Orchard.ContentManagement;
+using Orchard.Environment.Extensions;
+using Orchard.Mvc.Filters;
+using Orchard.Roles.Models;
+using Orchard.Security;
+
+namespace Orchard.OutputCache.Filters {
```

Comment:
I'd rather see these files in the Roles module so they don't take a direct dependency on the module. Or try to get the roles without the UserRolesPart.

Reply:
Looks to me all the records that would let us compute Roles and Permissions from the User are in Orchard.Roles. Even going around UserRolesPart, the "clean" way would probably be using IRepository<UserRolesPartRecord>, and possibly Join that with other stuff from the Orchard.Roles module if we want to get the Permissions (see the comment about hashing a list of those).
Orchard.Roles is one of those "Core" features that cannot be disabled from the UI. 
On the other hand, if we moved this feature in Orchard.Roles, that would create a dependency there from Orchard.OutputCache, where the ICachingEventHandler is defined.

Reply:
Adding to that:
The functionality would live in its own feature either way, so enabling Orchard.OutputCache or Orchard.Roles or both would not cause it to be enabled by default.